examples/CSharp/OpenGL Tutorials/Tutorial 1.3 - Textures/Program.cs

```diff
@@ -0,0 +1,306 @@
+
+            // Check to make sure that the shader has successfully compiled.
+            _gl.GetShader(vertexShader, ShaderParameterName.CompileStatus, out int vStatus);
+            if (vStatus != (int) GLEnum.True)
+                throw new Exception("Vertex shader failed to compile: " + _gl.GetShaderInfoLog(vertexShader));
+
+            // Repeat this process for the fragment shader.
+            uint fragmentShader = _gl.CreateShader(ShaderType.FragmentShader);
+            _gl.ShaderSource(fragmentShader, fragmentCode);
+
+            _gl.CompileShader(fragmentShader);
+
+            _gl.GetShader(fragmentShader, ShaderParameterName.CompileStatus, out int fStatus);
+            if (fStatus != (int) GLEnum.True)
+                throw new Exception("Fragment shader failed to compile: " + _gl.GetShaderInfoLog(fragmentShader));
+
+            // Create our shader program, and attach the vertex & fragment shaders.
+            _program = _gl.CreateProgram();
+
+            _gl.AttachShader(_program, vertexShader);
+            _gl.AttachShader(_program, fragmentShader);
+
+            // Attempt to "link" the program together.
+            _gl.LinkProgram(_program);
+
+            // Similar to shader compilation, check to make sure that the shader program has linked properly.
+            _gl.GetProgram(_program, ProgramPropertyARB.LinkStatus, out int lStatus);
+            if (lStatus != (int) GLEnum.True)
+                throw new Exception("Program failed to link: " + _gl.GetProgramInfoLog(_program));
+
+            // Detach and delete our shaders. Once a program is linked, we no longer need the individual shader objects.
+            _gl.DetachShader(_program, vertexShader);
+            _gl.DetachShader(_program, fragmentShader);
+            _gl.DeleteShader(vertexShader);
+            _gl.DeleteShader(fragmentShader);
+
+            // Our stride constant. The stride must be in bytes, so we take the first attribute (a vec3), multiply it
+            // by the size in bytes of a float, and then take our second attribute (a vec2), and do the same.
+            const uint stride = (3 * sizeof(float)) + (2 * sizeof(float));
+
+            // Enable the "aPosition" attribute in our vertex shader, providing its size and stride too.
+            const uint positionLoc = 0;
+            _gl.EnableVertexAttribArray(positionLoc);
+            _gl.VertexAttribPointer(positionLoc, 3, VertexAttribPointerType.Float, false, stride, (void*) 0);
+
+            // Now we need to enable our texture coordinates! We've defined that as location 1 so that's what we'll use
+            // here. The code is very similar to above, but you must make sure you set its offset to the **size in bytes**
+            // of the attribute before.
+            const uint textureLoc = 1;
+            _gl.EnableVertexAttribArray(textureLoc);
+            _gl.VertexAttribPointer(textureLoc, 2, VertexAttribPointerType.Float, false, stride, (void*) (3 * sizeof(float)));
```

Comment:
The comments here are conceptually wrong; you aren't enabling anything on the shader, but rather on the _vertex array_. This code is part of the VAO setup.

I'd actually recommend you put this before the shader setup (line ~89) to clarify this. I think it's also important to clarify what the `VertexAttribPointer` functions do.

These functions alter the currently bound VAO and specify what vertex attribute data they will send to a given attribute location. In this case, we're telling it to:
- Enable its attribute number 0 (done by EnableVertexAttribArray), otherwise all data read will be zeroed (or garbage? I'm not sure)
- Tell it that attribute number 0 will consist of 3 floats that should not be normalized, and that it should start reading this at an offset `offset=0` from the currently bound VBO (bound to ArrayBuffer), and each piece of data is separated by `stride` bytes (all done by VertexAttribPointer).

It's important to note that `VertexAttribPointer` alters _the currently bound VAO_, telling the VAO it has to use _the currently bound VBO_.

Reply:
Oof that's what sleep deprivation does I guess - I meant to put "vertex array". I'll add some extra comments too. Perhaps this should be added to the website tutorial as well?

Reply:
If you wish to, feel free to do so

Reply:
I'll look into adding this when I write the next set of tutorials for the website.